// @blueprintjs/popover2/lib/esnext/menuItem2.js
/*
 * Copyright 2022 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import classNames from "classnames";
import * as React from "react";
import { AbstractPureComponent2, Classes as CoreClasses, DISPLAYNAME_PREFIX, Icon, Menu, Text, } from "@blueprintjs/core";
import * as Classes from "./classes";
import { Popover2 } from "./popover2";
export class MenuItem2 extends AbstractPureComponent2 {
    static defaultProps = {
        active: false,
        disabled: false,
        multiline: false,
        popoverProps: {},
        selected: undefined,
        shouldDismissPopover: true,
        text: "",
    };
    static displayName = `${DISPLAYNAME_PREFIX}.MenuItem2`;
    render() {
        const { active, className, children, disabled, intent, labelClassName, labelElement, multiline, popoverProps, roleStructure = "menuitem", selected, shouldDismissPopover, submenuProps, text, textClassName, tagName = "a", htmlTitle, ...htmlProps } = this.props;
        const [liRole, targetRole, icon, ariaSelected] = roleStructure === "listoption"
            ? // "listoption": parent has listbox role, or is a <select>
                [
                    "option",
                    undefined,
                    this.props.icon ?? (selected === undefined ? undefined : selected ? "small-tick" : "blank"),
                    Boolean(selected), // aria-selected prop
                ]
            : // "menuitem": parent has menu role
                [
                    "none",
                    "menuitem",
                    this.props.icon,
                    undefined, // don't set aria-selected prop
                ];
        const hasIcon = icon != null;
        const hasSubmenu = children != null;
        const intentClass = CoreClasses.intentClass(intent);
        const anchorClasses = classNames(CoreClasses.MENU_ITEM, intentClass, {
            [CoreClasses.ACTIVE]: active,
            [CoreClasses.DISABLED]: disabled,
            // prevent popover from closing when clicking on submenu trigger or disabled item
            [Classes.POPOVER2_DISMISS]: shouldDismissPopover && !disabled && !hasSubmenu,
            [CoreClasses.SELECTED]: active && intentClass === undefined,
        }, className);
        const target = React.createElement(tagName, {
            role: targetRole,
            tabIndex: 0,
            ...htmlProps,
            ...(disabled ? DISABLED_PROPS : {}),
            className: anchorClasses,
        }, hasIcon ? (
        // wrap icon in a <span> in case `icon` is a custom element rather than a built-in icon identifier,
        // so that we always render this class
        React.createElement("span", { className: CoreClasses.MENU_ITEM_ICON },
            React.createElement(Icon, { icon: icon, "aria-hidden": true, tabIndex: -1 }))) : undefined, React.createElement(Text, { className: classNames(CoreClasses.FILL, textClassName), ellipsize: !multiline, title: htmlTitle }, text), this.maybeRenderLabel(labelElement), hasSubmenu ? React.createElement(Icon, { className: CoreClasses.MENU_SUBMENU_ICON, icon: "caret-right" }) : undefined);
        const liClasses = classNames({ [CoreClasses.MENU_SUBMENU]: hasSubmenu });
        return (React.createElement("li", { className: liClasses, role: liRole, "aria-selected": ariaSelected }, this.maybeRenderPopover(target, children)));
    }
    maybeRenderLabel(labelElement) {
        const { label, labelClassName } = this.props;
        if (label == null && labelElement == null) {
            return null;
        }
        return (React.createElement("span", { className: classNames(CoreClasses.MENU_ITEM_LABEL, labelClassName) },
            label,
            labelElement));
    }
    maybeRenderPopover(target, children) {
        if (children == null) {
            return target;
        }
        const { disabled, popoverProps, submenuProps } = this.props;
        return (React.createElement(Popover2, { autoFocus: false, captureDismiss: false, disabled: disabled, enforceFocus: false, hoverCloseDelay: 0, interactionKind: "hover", modifiers: SUBMENU_POPOVER_MODIFIERS, placement: "right-start", usePortal: false, ...popoverProps, content: React.createElement(Menu, { ...submenuProps }, children), minimal: true, popoverClassName: classNames(CoreClasses.MENU_SUBMENU, popoverProps?.popoverClassName) }, target));
    }
}
const SUBMENU_POPOVER_MODIFIERS = {
    // 20px padding - scrollbar width + a bit
    flip: { options: { rootBoundary: "viewport", padding: 20 }, enabled: true },
    // shift popover up 5px so MenuItems align
    offset: { options: { offset: [-5, 0] }, enabled: true },
    preventOverflow: { options: { rootBoundary: "viewport", padding: 20 }, enabled: true },
};
// props to ignore when disabled
const DISABLED_PROPS = {
    href: undefined,
    onClick: undefined,
    onMouseDown: undefined,
    onMouseEnter: undefined,
    onMouseLeave: undefined,
    tabIndex: -1,
};
